File: QE/R/Uniprot_GO.R
options(warn =-1)
if (!require("pacman")){
  install.packages("pacman",dep=TRUE, repos='https://mirrors.tuna.tsinghua.edu.cn/CRAN/')
  if(!require("pacman")){
  stop("安装失败")
  }
}

p_load(dplyr, foreach, doSNOW, xml2, openxlsx, doParallel)

# suppressMessages(library(dplyr))
# suppressMessages(library(foreach))
# suppressMessages(library(doSNOW))
# # library(XML)
# library(xml2)
# library(openxlsx)
# suppressMessages(library(doParallel))

args <- commandArgs(T)
path <- as.character(args[1])
thread <- as.numeric(args[2])
setwd(path)


################# FUNCTION ######################
# proid <- "aaabbb"
# proid <- "A0A0G2JGX4"
# proid <- "A0A087WUF0"

# library(RCurl)
# library(xml2)
# library(dplyr)

get_info <- function(proid){
  library(xml2)
  library(dplyr)
  name <- paste0("http://www.uniprot.org/uniprot/", proid, ".xml")
  # result <- try(xmlTreeParse(name, useInternalNodes = TRUE), silent = T)
  result <- try(read_xml(name), silent = T)
  
  # res <- as_list(result)
  
  error <- 1
  if (!class(result)[1] == "xml_document"){
    while(any(class(result) == "try-error"& error == 1)){
      cat(paste0("The error protein ID is ", proid, ". If this error happens again, please check your network or protein ID !!!\n"))
      if (grepl("Failed to parse text", result[1])){
        error <- 0
      }else{
        result <- try(read_xml(name), silent = T)
      }
    }
  }
  
  if (any(class(result) == "try-error")){
    out <- c(ProteinID = proid, ProteinName = "-", BP = "-", MF = "-", CC = "-", GO_Term = "-")
  }else{
    # goid <- xpathSApply(result, "//g:dbReference[@type='GO']", xmlGetAttr, "id", namespaces = "g")
    # tmp <- xpathSApply(result, "//j:property [@type='term']", xmlGetAttr, "value", namespaces = "j")
    # category <- unlist(lapply(tmp, function(x){strsplit(x, ":")[[1]][1]}))
    # goterm <- unlist(lapply(tmp, function(x){strsplit(x, ":")[[1]][2]}))
    # proname <- xpathSApply(result, "//j:fullName", fun = xmlValue, namespaces = "j")[1]
    
    goid <- xml_find_all(result, ".//d1:dbReference[@type='GO']", ns = xml_ns(result)) %>% xml_attr("id")
    tmp <- xml_find_all(result, ".//d1:property[@type='term']", ns = xml_ns(result)) %>% xml_attr("value")
    category <- unlist(lapply(tmp, function(x){strsplit(x, ":")[[1]][1]}))
    goterm <- unlist(lapply(tmp, function(x){strsplit(x, ":")[[1]][2]}))
    proname <- xml_find_all(result, ".//d1:fullName", ns = xml_ns(result)) %>% xml_text() %>% head(1)
    
    if ("P" %in% category){
      bp <- paste0(paste(goterm[which(category == "P")], paste0("[", goid[which(category == "P")], "]")), collapse = ";")
    }else{
      bp <- "-"
    }
    
    if ("F" %in% category){
      mf <- paste0(paste(goterm[which(category == "F")], paste0("[", goid[which(category == "F")], "]")), collapse = ";")
    }else{
      mf <- "-"
    }
    
    if ("C" %in% category){
      cc <- paste0(paste(goterm[which(category == "C")], paste0("[", goid[which(category == "C")], "]")), collapse = ";")
    }else{
      cc <- "-"
    }
    
    all_go <- paste(c(bp[!is.na(bp)], mf[!is.na(mf)], cc[!is.na(cc)]), collapse = ";")
    
    if (length(goid) == 0){
      out <- c(ProteinID = proid, ProteinName = proname, BP = "-", MF = "-", CC = "-", GO_Term = "-")
    }else{
      out <- c(ProteinID = proid, ProteinName = proname, BP = bp, MF = mf, CC = cc, GO_Term = all_go)
    }
  }
  return(out)
}


dirs <- dir()
dirs <- dirs[!grepl("\\.", dirs)]

wb <- createWorkbook()
modifyBaseFont(wb, fontName = "Arial", fontSize = 10)

project_number <- 1

for(i in dirs){
  folder <- i
  file <- paste0(folder, "/diff_list.txt")
  
  uniprot_id <- read.table(file = file, header = F, stringsAsFactors = F) %>% as.matrix() %>% as.character()
  # 
  cl.cores <- detectCores()
  if (is.na(thread)){
    core <- 10
  }else{
    core <- thread
  }
  
  #  core <- 10
  cl <- makeCluster(core)
  registerDoSNOW(cl)
  pb <- txtProgressBar(min = 0, max = length(uniprot_id), style = 3)
  progress <- function(n) setTxtProgressBar(pb, n)
  opts <- list(progress = progress)
  
  system.time({
    result_df <- foreach(i = 1:length(uniprot_id), .combine = rbind, .options.snow = opts) %dopar% {
      df <- get_info(proid = uniprot_id[i])
      
      Sys.sleep(0.2)
      return(df)
    }
    close(pb)
    stopCluster(cl)
  })
  
  # for(i in 1:length(uniprot_id)){
  #   df <- get_info(proid = uniprot_id[i])
  #   result_df <- rbind(result_df, df)
  #   
  #   Sys.sleep(0.5)
  #   setTxtProgressBar(pb, i)
  # }
  
  print(paste0("The processs of ", i, " is finished !!!"))
  
  addWorksheet(wb, sheetName = folder)
  header_style <- createStyle(textDecoration = "bold", halign = "left")
  addStyle(wb, sheet = folder, rows = 1, cols = 1:ncol(result_df), style = header_style)
  
  setColWidths(wb, sheet = folder, cols = 2, widths = 15)
  writeData(wb, sheet = folder, x = result_df)
}

saveWorkbook(wb, "GO.xlsx", overwrite = T)

File: QE/uniprot_go.py
import os
import re
import time
import socket
from tqdm import tqdm
from queue import Queue
from functools import wraps
from threading import Thread
from openpyxl import Workbook
from urllib.request import urlopen
import xml.etree.ElementTree as ET
# from multiprocessing import Pool

def fn_timer(function):
    @wraps(function)
    def function_timer(*args, **kwargs):
        t0 = time.time()
        result = function(*args, **kwargs)
        t1 = time.time()
        print(f"Total time running: {str(int(t1-t0))} seconds")
        return result
    return function_timer


class Spider():
    def __init__(self, path):
        self.path = path
        self.proid = Queue()
        self.data = list()
        self.thread_num = 20

    def produce_xml(self, dir):
        url = 'http://www.uniprot.org/uniprot/{}.xml'
        file = os.path.join(dir, 'diff_list.txt')
        diff_list = open(file)
        for j in diff_list:
            xml = url.format(j.strip())
            self.proid.put(xml)
        diff_list.close()

    def get_info(self):
        while not self.proid.empty():
            xml = self.proid.get()
            proid = re.split('[/.]', xml)[-2]
            try:
                response = urlopen(xml)
                tree = ET.parse(response)
                root = tree.getroot()
                goid = [child.attrib.get('id') for child in root.iter(tag='{http://uniprot.org/uniprot}dbReference') if child.attrib.get('type') == 'GO']
                tmp = [child.attrib.get('value') for child in root.iter(tag='{http://uniprot.org/uniprot}property') if child.attrib.get('type') == 'term']
                category = [i.split(':')[0] for i in tmp]
                goterm = [i.split(':')[1] for i in tmp]
                proname = ''.join([child.text for child in root.iter(tag='{http://uniprot.org/uniprot}fullName')][0])
                geneid = ''.join([child.attrib.get('id') for child in root.iter(tag='{http://uniprot.org/uniprot}dbReference') if child.attrib.get('type') == 'GeneID'])
                genename = ''.join([child.text for child in root.iter(tag='{http://uniprot.org/uniprot}name') if child.attrib.get('type') == 'primary'])

                # proname = None if len(proname) == 0 else proname
                # geneid = None if len(geneid) == 0 else geneid
                # genename = None if len(genename) == 0 else genename

                bp = set([f'{goterm[k]} [{goid[k]}]' if v == 'P' else '-' for k, v in enumerate(category)])
                mf = set([f'{goterm[k]} [{goid[k]}]' if v == 'F' else '-' for k, v in enumerate(category)])
                cc = set([f'{goterm[k]} [{goid[k]}]' if v == 'C' else '-' for k, v in enumerate(category)])
                all_go = []
                all_go.extend(bp)
                all_go.extend(mf)
                all_go.extend(cc)

                if len(goid) == 0:
                    out = f'{proid}\t{proname}\t{geneid}\t{genename}\t-\t-\t-\t-'.split('\t')
                else:
                    bp = ';'.join(bp)
                    mf = ';'.join(mf)
                    cc = ';'.join(cc)
                    all_go = ';'.join(all_go)
                    out = f'{proid}\t{proname}\t{geneid}\t{genename}\t{bp}\t{mf}\t{cc}\t{all_go}'.split('\t')
                self.data.append(out)

                response.close()
            except Exception as e:
                out = f'{proid}\t-\t-\t-\t-\t-\t-\t-'.split('\t')
                self.data.append(out)
                print(e)
    # print(get_info('A1YZ34'))
    @fn_timer
    def main(self):
        socket.setdefaulttimeout(20)
        wb = Workbook()
        i = 0
        for d in tqdm(os.listdir(self.path)):
            dir = os.path.join(self.path, d)
            if os.path.isdir(dir):
                self.produce_xml(dir)
                threads = []
                for _ in range(self.thread_num):
                    t = Thread(target=self.get_info)
                    time.sleep(1)
                    t.start()
                    threads.append(t)
                for t in threads:
                    t.join()
                ws = wb.create_sheet(d, i)
                i += 1
                ws.append(['ProteinID', 'ProteinName', 'GeneID', 'GeneName', 'BP', 'MF', 'CC', 'GO_Term'])
                for j in self.data:
                    # print(j)
                    ws.append(j)
                self.data.clear()

        wb.save(os.path.join(self.path, 'GO.xlsx'))

# if __name__=='__main__':
#     path = 'E:\\GUI\\QE_exe\\Data\\R20190901243'
#     # start = time.time()
#     Spider(path).main()
#     # end = time.time()
#     # print(end - start)
